### PR TITLE
Added Nadam

### PR DESCRIPTION
**Pull request by greatsharma**
Implemented nadam optimizer. I followed [this](https://ruder.io/optimizing-gradient-descent/index.html#nadam) paper mentioned in  #173

### CODE REVIEW

File: creme/optim/nadam.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        lr = self.learning_rate * \
+            (1 - self.beta_2 ** (self.n_iterations + 1)) ** .5
+        # Correct bias for `m`
+        lr /= (1 - self.beta_1 ** (self.n_iterations + 1))
+
+        for i, gi in g.items():
+            self.m[i] = self.beta_1 * self.m[i] + (1 - self.beta_1) * gi
+            self.v[i] = self.beta_2 * self.v[i] + (1 - self.beta_2) * gi ** 2
+
+            w[i] -= lr * (self.beta_1*self.m[i] + (1-self.beta_1)*gi*(1-np.power(
```
Comment (AdilZouitine) [inline creme/optim/nadam.py]: We try to be as least dependent on numpy as possible, I think it would be better to use the power operation available in the standard `math` library 😄

**Comment by MaxHalford** [comment]
I'm having a hard-type verifying if the implementation is correct. @VaysseRobin do you have any thoughts?

**Comment by greatsharma** [comment]
@MaxHalford where you think the implementation could be wrong?

Also in adam the default value for `lr` is `0.1` but in the mentioned original paper for adam the default value is `0.001`. Shouldn't we change it?

File: creme/optim/nadam.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            (1 - self.beta_2 ** (self.n_iterations + 1)) ** .5
+        # Correct bias for `m`
+        lr /= (1 - self.beta_1 ** (self.n_iterations + 1))
+
+        for i, gi in g.items():
+            self.m[i] = self.beta_1 * self.m[i] + (1 - self.beta_1) * gi
+            self.v[i] = self.beta_2 * self.v[i] + (1 - self.beta_2) * gi ** 2
+
+            w[i] -= lr * (self.beta_1*self.m[i] + (1-self.beta_1)*gi*(1-math.pow(
+                self.beta_1, self.n_iterations+1))) / (self.v[i] ** .5 + self.eps)
```
Comment (VaysseRobin) [inline creme/optim/nadam.py]: The implementation is almost correct but I see two problems (correct me if I am wrong), the first one is here `gi*(1-math.pow(
                self.beta_1, self.n_iterations+1))`, it should be a division instead of a multiplication between `gi` et the `(1-math.pow(...)`

The second one is a little bit tricky but I will try to explain you with some math.

Comment (VaysseRobin) [reply creme/optim/nadam.py]: The issue is because you do the bias correction in the learning rate which is usaually not a problem but in the case of NADAM, it is:

NADAM

So you should do the bias correction at each step.

Comment (greatsharma) [reply creme/optim/nadam.py]: > The implementation is almost correct but I see two problems (correct me if I am wrong), the first one is here `gi*(1-math.pow( self.beta_1, self.n_iterations+1))`, it should be a division instead of a multiplication between `gi` et the `(1-math.pow(...)`
> 
Thank you @VaysseRobin for this, it was a typo and lucky it worked well even with this. I changed it now

Comment (greatsharma) [reply creme/optim/nadam.py]: > The issue is because you integrate bias correction in the learning rate which is usaually not a problem but in the case of NADAM, it is:
> 
> NADAM
> 
> So you should do the bias correction at each step.

@VaysseRobin I changed it now. Thankyou :)

**Comment by greatsharma** [comment]
> Also in adam the default value for `lr` is `0.1` but in the mentioned original paper for adam the default value is `0.001`. Shouldn't we change it?

@VaysseRobin what's your point on this?

File: creme/optim/nadam.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    def __init__(self, lr=0.1, beta_1=0.9, beta_2=0.999, eps=1e-8):
+        super().__init__(lr)
+        self.beta_1 = beta_1
+        self.beta_2 = beta_2
+        self.eps = eps
+        self.m = collections.defaultdict(float)
+        self.v = collections.defaultdict(float)
+        self.m_hat = collections.defaultdict(float)
+        self.v_hat = collections.defaultdict(float)
```
Comment (VaysseRobin) [inline creme/optim/nadam.py]: You don't need to store history of `m_hat` and `v_hat` since the update of `m` and  `v` does not rely on them. So you should just use local variables in `_update_after_pred`.

**Comment by VaysseRobin** [comment]
> > Also in adam the default value for `lr` is `0.1` but in the mentioned original paper for adam the default value is `0.001`. Shouldn't we change it?
> 
> @VaysseRobin what's your point on this?

I think that it does not really matter, but yes it could be good to match the paper.

File: creme/optim/nadam.py (hunk 3 above)
Comment (greatsharma) [reply creme/optim/nadam.py]: I changed it now

**Comment by VaysseRobin** [comment]
It seems good to me, nice job ! @MaxHalford I think you can merge it.

**Comment by greatsharma** [comment]
> > > Also in adam the default value for `lr` is `0.1` but in the mentioned original paper for adam the default value is `0.001`. Shouldn't we change it?
> > 
> > 
> > @VaysseRobin what's your point on this?
> 
> I think that it does not really matter, but yes it could be good to match the paper.

@MaxHalford what you think?

**Comment by MaxHalford** [comment]
@greatsharma I think it doesn't really matter. What we will do one day is to define a large set of benchmarks, we will then benchmark each optimizer and set the parameters that are the best overall.
